contracts/Smt.metasol

```diff
@@ -1,27 +1,36 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
-library SMT{{hashname}} {
+library SMT{{subcontractname}} {
     bytes32 constant internal EMPTY_SUBTREE = 0x00;
 
     struct MerkleTree {
         uint8 depth;
+        uint8 maxDepth;
+        {{ if SET }}
+        uint256 firstEmptySlot;
+        {{ else }}
+        {{ endif }}
         mapping(uint256 => bytes32) emptySubtreeCache;
         // tree level => index inside level => element hash
         mapping(uint256 => mapping(uint256 => bytes32)) nodes;
         mapping(uint256 => bytes) elementData;
     }
 
-    using SMT{{hashname}} for MerkleTree;
+    using SMT{{subcontractname}} for MerkleTree;
 
-    function initialize(MerkleTree storage tree, uint8 _depth) internal {
+    function _initialize(MerkleTree storage tree, uint8 _depth, uint8 _maxDepth) private {
         require(tree.depth == 0, "Cannot initialize twice");
         require(_depth > 0, "Depth must be non-zero");
         require(_depth < 255, "Depth must be < 255");
+        require(_maxDepth > 0, "Max depth must be non-zero");
+        require(_maxDepth < 255, "Max depth must be < 255");
```

Comment:
IMO, we need to add a check that `_depth` <= `_maxDepth`, I know that `_initialize` calls with parameters that satisfy this invariant, but it is better to set require to prevent bugs in future.

Also we can have only 3 `require`s instead of 4. 

```suggestion
        require(_depth > 0, "Depth must be non-zero");
        require(_maxDepth < 255, "Max depth must be < 255");
        require(_depth <= _maxDepth, "Depth must be <= Max depth");
```